rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,419 @@
+            symcrypt_sys::SymCryptHmacSha256Result(&mut self.inner.state, result.as_mut_ptr());
+        }
+        result
+    }
+
+    fn copy(&self) -> Box<Self> {
+        let mut new_state = HmacSha256State {
+            inner: Box::pin(HmacSha256Inner {
+                state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE::default(),
+                expanded_key: Arc::clone(&self.inner.expanded_key),
+            }),
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptHmacSha256StateCopy(
+                &self.inner.state,
+                &*self.inner.expanded_key.0,
+                &mut new_state.inner.state,
+            );
         }
+        Box::new(new_state)
     }
 }
 
 impl Drop for HmacSha256State {
     fn drop(&mut self) {
         unsafe {
+            // SAFETY: FFI calls
             symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.state) as *mut c_void,
-                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
-            );
-
-            symcrypt_sys::SymCryptWipe(
-                ptr::addr_of_mut!(self.expanded_key) as *mut c_void,
-                mem::size_of_val(&mut self.expanded_key) as symcrypt_sys::SIZE_T,
+                ptr::addr_of_mut!(self.inner.state) as *mut c_void,
+                mem::size_of_val(&mut self.inner.state) as symcrypt_sys::SIZE_T,
             );
         }
     }
 }
 
-pub fn hmac_sha256(key: &[u8], data: &[u8], result: &mut [u8; SHA256_HMAC_RESULT_SIZE]) {
+/// Stateless Hmac functions for Sha256
+pub fn hmac_sha256(
+    key: &[u8],
+    data: &[u8],
+    result: &mut [u8; SHA256_HMAC_RESULT_SIZE],
+) -> Result<(), SymCryptError> {
     unsafe {
+        // SAFETY: FFI calls
         let mut expanded_key = symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default();
-        symcrypt_sys::SymCryptHmacSha256ExpandKey(
+        match symcrypt_sys::SymCryptHmacSha256ExpandKey(
             &mut expanded_key,
             key.as_ptr(),
             key.len() as symcrypt_sys::SIZE_T,
-        );
-        symcrypt_sys::SymCryptHmacSha256(
-            &mut expanded_key,
-            data.as_ptr(),
-            data.len() as symcrypt_sys::SIZE_T,
-            result.as_mut_ptr(),
-        );
-        symcrypt_sys::SymCryptWipe(
-            ptr::addr_of_mut!(expanded_key) as *mut c_void,
-            mem::size_of_val(&mut expanded_key) as symcrypt_sys::SIZE_T,
-        );
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                symcrypt_sys::SymCryptHmacSha256(
+                    &mut expanded_key,
+                    data.as_ptr(),
+                    data.len() as symcrypt_sys::SIZE_T,
+                    result.as_mut_ptr(),
+                );
+
+                symcrypt_sys::SymCryptWipe(
+                    ptr::addr_of_mut!(expanded_key) as *mut c_void,
+                    mem::size_of_val(&mut expanded_key) as symcrypt_sys::SIZE_T,
```

Comment:
This can be removed when you wrap via HmacSha256ExpandedKey

Reply:
Removed